aopy/preproc/bmi3d.py

```diff
@@ -481,33 +481,34 @@ def find_laser_stim_times(laser_event_times, laser_event_widths, laser_event_pow
-    laser_on_times = np.mean([rising, falling], axis=0)
-    laser_on_samples = (laser_on_times * ds_fs).astype(int)
-    laser_sensor_powers = (ds_data[laser_on_samples])*sensor_voltsperbit
+    laser_sensor_off_times = falling
     
-    # Normalize sensor power to the highest power, then multiply by the highest trial power
-    laser_sensor_powers /= np.max(laser_sensor_powers)
-    laser_sensor_powers *= np.max(laser_event_powers)
-
     # Check that the sensor measurements make sense, otherwise return the sync event versions
     corrected_times, corrected_idx = find_measured_event_times(laser_event_times, laser_sensor_times, search_radius, return_idx=True)
     missing_times = np.isnan(corrected_times)
     if np.any(missing_times):
         warnings.warn(f"{np.count_nonzero(missing_times)} unmeasured laser timestamps")
         corrected_times[missing_times] = laser_event_times[missing_times]
+    corrected_off_times, corrected_off_idx = find_measured_event_times(laser_event_times+laser_event_widths, laser_sensor_off_times, search_radius, return_idx=True)
+    missing_off_times = np.isnan(corrected_off_times)
+    if np.any(missing_off_times):
+        warnings.warn(f"{np.count_nonzero(missing_times)} unmeasured laser offsets")
+        corrected_off_times[missing_off_times] = laser_event_times[missing_off_times] + laser_event_widths[missing_off_times]
+    times_not_found = np.logical_or(np.isnan(corrected_idx), np.isnan(corrected_off_idx))
+
+    # Now calculate the widths and powers based on corrected times
+    laser_sensor_widths = corrected_off_times - corrected_times
+    laser_on_times = np.mean([corrected_off_times, corrected_times], axis=0)
+    laser_on_samples = (laser_on_times * ds_fs).astype(int)
+    laser_sensor_powers = (ds_data[laser_on_samples])*sensor_voltsperbit
     
-    # Adjust the sensor measurements to remove any that were missing
-    times_not_found = np.isnan(corrected_idx)
-    valid_idx = corrected_idx[~times_not_found].astype(int)
-    adjusted_widths = corrected_idx.copy()
-    adjusted_widths[~np.isnan(corrected_idx)] = laser_sensor_widths[valid_idx]
-    adjusted_powers = corrected_idx.copy()
-    adjusted_powers[~np.isnan(corrected_idx)] = laser_sensor_powers[valid_idx]
+    # Normalize sensor power to the highest power, then multiply by the highest trial power
+    laser_sensor_powers /= np.max(laser_sensor_powers)
+    laser_sensor_powers *= np.max(laser_event_powers)
 
     # Correct the widths and powers with the given thresholds
-    corrected_widths, widths_above_thr = validate_measurements(laser_event_widths, adjusted_widths, thr_width)
-    corrected_powers, powers_above_thr = validate_measurements(laser_event_powers, adjusted_powers, thr_power)
+    corrected_widths, widths_above_thr = validate_measurements(laser_event_widths, laser_sensor_widths, thr_width)
+    corrected_powers, powers_above_thr = validate_measurements(laser_event_powers, laser_sensor_powers, thr_power)
```

Comment:
If i'm understanding the code correctly this assumes that np.max(laser_event_powers) is the same as np.max(laser_sensor_powers). Is this okay?

Reply:
it is supposed to *force* them to be the same. i figure the sensor measurement is meaningless anyway, so just scale it to match the given event power